apps/cload/cload.c

```diff
@@ -0,0 +1,263 @@
+
+
+typedef unsigned char uint8_t;
+typedef unsigned long uint32_t;
+typedef unsigned long size_t;
+typedef long long int64_t;
+typedef unsigned long long uint64_t;
+
+int cload_main(int cmd, uint32_t *io, char *ram, char *fb);
+
+// Don't call it main, it goes into different section
+// Make it assembly jump, so that if optimizer reorders stuff, this is still first
+__attribute__((naked, used, section(".text.startup")))  int cload_entry(int cmd, uint32_t *io, char *ram, char *fb) {
+    __asm__ volatile (
+	    "b cload_main");
+}
+
+static char *dmesg;
+
+void putchar(char v)
+{
+	*dmesg++ = v;
+}
+
+void puts(const char *s) {
+    while (*s) {
+        putchar(*s++);
+    }
+    putchar('\n');
+}
+
+void reverse(char *str, int len) {
+    int i = 0, j = len - 1;
+    while (i < j) {
+        char tmp = str[i];
+        str[i++] = str[j];
+        str[j--] = tmp;
+    }
+}
+
+void itoa(int value, char *str) {
+    int i = 0;
+    int is_negative = 0;
+
+    if (value == 0) {
+        str[i++] = '0';
+        str[i] = '\0';
+        return;
+    }
+
+    if (value < 0) {
+        is_negative = 1;
+        value = -value;
+    }
+
+    while (value != 0) {
+        int digit = value % 10;
+        str[i++] = digit + '0';
+        value /= 10;
+    }
+
+    if (is_negative)
+        str[i++] = '-';
+
+    str[i] = '\0';
+    reverse(str, i);
+}
+
+void *memset(void *dest, int val, size_t len) {
+    unsigned char *ptr = dest;
+    while (len-- > 0) {
+        *ptr++ = (unsigned char)val;
+    }
+    return dest;
+}
+
+typedef struct {
+    uint32_t sp;   // offset 0 -- aka r13
+    uint32_t lr;   // offset 4 -- aka r14
+    uint32_t r4;   // offset 8
+    uint32_t r5;   // ...
+    uint32_t r6;
+    uint32_t r7;
+    uint32_t r8;
+    uint32_t r9;
+    uint32_t r10;
+    uint32_t r11;  // offset 40
+} context_t;
+
+__attribute__((naked)) void switch_stack(context_t *old_ctx, context_t *new_ctx) {
+    __asm__ volatile (
+        // Save old context
+        "str sp, [r0, #0]       \n"
+        "str lr, [r0, #4]       \n"
+        "str r4, [r0, #8]       \n"
+        "str r5, [r0, #12]      \n"
+        "str r6, [r0, #16]      \n"
+        "str r7, [r0, #20]      \n"
+        "str r8, [r0, #24]      \n"
+        "str r9, [r0, #28]      \n"
+        "str r10, [r0, #32]     \n"
+        "str r11, [r0, #36]     \n"
+
+        // Load new context
+        "ldr sp, [r1, #0]       \n"
+        "ldr lr, [r1, #4]       \n"
+        "ldr r4, [r1, #8]       \n"
+        "ldr r5, [r1, #12]      \n"
+        "ldr r6, [r1, #16]      \n"
+        "ldr r7, [r1, #20]      \n"
+        "ldr r8, [r1, #24]      \n"
+        "ldr r9, [r1, #28]      \n"
+        "ldr r10, [r1, #32]     \n"
+        "ldr r11, [r1, #36]     \n"
+
+        "bx lr                  \n"
+    );
+}
+
+volatile char welcome[] = "Hello\nwatch\n:-)";
```

Comment:
If you make these (and all functions except public ones) `static` you'll get better loads/stores and potentially less memory use with the GOT depending on how the linker fares